components/eamxx/src/physics/mam/mam_aerosol_optics_read_tables.hpp

```diff
@@ -0,0 +1,366 @@
+  layouts.emplace("refindex_im_lw", scalar_refindex_im_lw_layout);
+  layouts.emplace("refindex_real_sw", scalar_refindex_real_sw_layout);
+  layouts.emplace("refindex_im_sw", scalar_refindex_im_sw_layout);
+  layouts.emplace("absplw", scalar_absplw_layout);
+  layouts.emplace("asmpsw", scalar_asmpsw_layout);
+  layouts.emplace("extpsw", scalar_asmpsw_layout);
+  layouts.emplace("abspsw", scalar_asmpsw_layout);
+}
+// KOKKOS_INLINE_FUNCTION
+inline void read_rrtmg_table(
+    const std::string &table_filename, const int imode,
+    ekat::ParameterList &params,
+    const std::shared_ptr<const AbstractGrid> &grid,
+    const std::map<std::string, view_1d_host> &host_views_1d,
+    const std::map<std::string, FieldLayout> &layouts,
+    const AerosolOpticsHostData &aerosol_optics_host_data,
+    const AerosolOpticsDeviceData &aerosol_optics_device_data) {
+  constexpr int refindex_real = mam4::modal_aer_opt::refindex_real;
+  constexpr int refindex_im   = mam4::modal_aer_opt::refindex_im;
+  constexpr int coef_number   = mam4::modal_aer_opt::coef_number;
+
+  using view_3d_host = typename KT::view_3d<Real>::HostMirror;
+
+  // temp views:
+  view_3d_host temp_lw_3d_host("temp_absplw_host", coef_number, refindex_real,
+                               refindex_im);
+
+  params.set("Filename", table_filename);
+  AtmosphereInput rrtmg(params, grid, host_views_1d, layouts);
+  rrtmg.read_variables();
+  rrtmg.finalize();
+
+  // copy data from host to device for mode 1
+  int d1 = imode;
+  for(int d3 = 0; d3 < nswbands; ++d3) {
+    auto real_host_d3 = ekat::subview(
+        aerosol_optics_host_data.refindex_real_sw_host, d3);
+    Kokkos::deep_copy(aerosol_optics_device_data.refrtabsw[d1][d3],
+                      real_host_d3);
+    auto im_host_d3 = ekat::subview(
+        aerosol_optics_host_data.refindex_im_sw_host, d3);
+    Kokkos::deep_copy(aerosol_optics_device_data.refitabsw[d1][d3], im_host_d3);
+  }  // d3
+
+  for(int d3 = 0; d3 < nlwbands; ++d3) {
+    auto real_host_d3 = ekat::subview(
+        aerosol_optics_host_data.refindex_real_lw_host, d3);
+    Kokkos::deep_copy(aerosol_optics_device_data.refrtablw[d1][d3],
+                      real_host_d3);
+    auto im_host_d3 = ekat::subview(
+        aerosol_optics_host_data.refindex_im_lw_host, d3);
+    Kokkos::deep_copy(aerosol_optics_device_data.refitablw[d1][d3], im_host_d3);
+  }  // d3
+
+  // NOTE: we need to reorder dimensions in absplw
+  // netcfd : (lw_band, mode, refindex_im, refindex_real, coef_number)
+  // mam4xx : (mode, lw_band, coef_number, refindex_real, refindex_im )
+  // e3sm : (ntot_amode,coef_number,refindex_real,refindex_im,nlwbands)
+
+  for(int d5 = 0; d5 < nlwbands; ++d5) {
+    // reshape data:
+    for(int d2 = 0; d2 < coef_number; d2++)
+      for(int d3 = 0; d3 < refindex_real; d3++)
+        for(int d4 = 0; d4 < refindex_im; d4++)
+          temp_lw_3d_host(d2, d3, d4) =
+              aerosol_optics_host_data.absplw_host(d5, 0, d4, d3, d2);
```

Comment:
We cannot use  Kokkos::subview, because we must re-order dimensions in these tables. For example, 
temp_lw_3d_host has a layout <d2,d3,d4>, but aerosol_optics table has <d4, d3, d2>. I added a note before that loops. 
```

  // NOTE: we need to reorder dimensions in absplw
  // netcfd : (lw_band, mode, refindex_im, refindex_real, coef_number)
  // mam4xx : (mode, lw_band, coef_number, refindex_real, refindex_im ) 
```

Reply:
I added a `Kokkos::parallel_for` to reshape these tables.

Reply:
Ah, I missed that detail.